openedx/core/djangoapps/zendesk_proxy/utils.py

```diff
@@ -0,0 +1,67 @@
+"""
+Utility functions for zendesk interaction.
+"""
+import json
+import logging
+from urlparse import urljoin
+
+from django.conf import settings
+import requests
+
+log = logging.getLogger(__name__)
+
+
+def create_zendesk_ticket(requester_name, requester_email, subject, body, tags=None):
+    """
+    Create a Zendesk ticket via API.
+
+    Note that we do this differently in other locations (lms/djangoapps/commerce/signals.py and
+    common/djangoapps/util/views.py). Both of those callers use basic auth, and should be switched over to this oauth
+    implementation once the immediate pressures of zendesk_proxy are resolved.
+    """
+    if not (settings.ZENDESK_URL and settings.ZENDESK_OAUTH_ACCESS_TOKEN):
+        log.debug('Zendesk is not configured. Cannot create a ticket.')
+        return
+
+    # Remove duplicates from tags list
+    tags = list(set(tags))
+
+    data = {
+        'ticket': {
+            'requester': {
+                'name': requester_name,
+                'email': requester_email
+            },
+            'subject': subject,
+            'comment': {'body': body},
+            'tags': tags
+        }
+    }
+
+    # Encode the data to create a JSON payload
+    payload = json.dumps(data)
+
+    # Set the request parameters
+    url = urljoin(settings.ZENDESK_URL, '/api/v2/tickets.json')
+    headers = {
+        'content-type': 'application/json',
+        'Authorization': "Bearer {}".format(settings.ZENDESK_OAUTH_ACCESS_TOKEN),
+    }
+
+    try:
+        response = requests.post(url, data=payload, headers=headers)
+
+        # Check for HTTP codes other than 201 (Created)
+        if response.status_code != 201:
+            log.error(
+                'zendesk_proxy unexpected HTTP response: {} - {}. No ticket created for payload {}'.format(
+                    response.status_code,
+                    response.content,
+                    payload
+                )
+            )
+        else:
+            log.debug('Successfully created ticket for {}'.format(requester_email))
+    except Exception:  # pylint: disable=broad-except
+        log.exception('zendesk_proxy exception: Failed to create ticket. Request data: {}'.format(payload))
```

Comment:
This one will need to be given a splunk alert so that we can send emails in the event zendesk isn't working for whatever reason. This will still be protected by the rate limiting logic, as it's called by the view code (rate limiting bounces execution before the view method is ever touched).